File: documentation/AllMetaheuristic/Nature-inspired/physics&chemistry-based/Thermal Exchange Optimization_re/TEO/TEO.m
% Thermal Exchange Optimization (TEO) Implementation

% Parameters
population_size = 50;        % Population size
num_variables = 10;          % Number of variables
max_iterations = 100;        % Maximum number of iterations
step_size = 0.1;             % Step size for movement
temperature_initial = 100;   % Initial temperature
temperature_final = 0.01;    % Final temperature
cooling_rate = 0.99;         % Cooling rate

% Initialization
current_solution = rand(1, num_variables); % Random initial solution
current_fitness = objective_function(current_solution);
best_solution = current_solution;
best_fitness = current_fitness;

% Main loop
temperature = temperature_initial;
for iter = 1:max_iterations
    % Generate a new solution by perturbing the current solution
    new_solution = current_solution + step_size * randn(1, num_variables);
    
    % Evaluate the new solution
    new_fitness = objective_function(new_solution);
    
    % Determine whether to accept the new solution
    if new_fitness < current_fitness || rand() < exp((current_fitness - new_fitness) / temperature)
        current_solution = new_solution;
        current_fitness = new_fitness;
    end
    
    % Update the best solution found so far
    if current_fitness < best_fitness
        best_solution = current_solution;
        best_fitness = current_fitness;
    end
    
    % Reduce temperature
    temperature = temperature * cooling_rate;
    
    % Check for convergence
    if temperature < temperature_final
        break;
    end
end

% Display best solution found
fprintf('Best solution: %s\n', mat2str(best_solution));
fprintf('Best fitness: %f\n', best_fitness);
% Objective function (example)
function z = objective_function(x)
    % Example: Sphere function
    z = sum(x.^2);
end

File: chekml/MetaheuristicOptimization/CIntegration_cffi2/TEO.c
#include "TEO.h"
#include "generaloptimizer.h"
#include <stdlib.h>
#include <time.h>
#include <math.h>

// Function to generate a random double between min and max
double rand_double(double min, double max);

// Initialize the current solution (randomly within bounds)
void initialize_solution(Optimizer *opt, double *current_solution, double *current_fitness, double (*objective_function)(double *)) {
    for (int j = 0; j < opt->dim; j++) {
        current_solution[j] = opt->bounds[2 * j] + (opt->bounds[2 * j + 1] - opt->bounds[2 * j]) * rand_double(0.0, 1.0);
    }
    *current_fitness = objective_function(current_solution);
    enforce_bound_constraints(opt);
    
    // Set best solution initially
    opt->best_solution.fitness = *current_fitness;
    for (int j = 0; j < opt->dim; j++) {
        opt->best_solution.position[j] = current_solution[j];
    }
}

// Perturb the current solution
void perturb_solution(Optimizer *opt, double *new_solution) {
    for (int j = 0; j < opt->dim; j++) {
        // Gaussian perturbation (approximated using uniform random numbers)
        double perturbation = STEP_SIZE * (rand_double(-1.0, 1.0) + rand_double(-1.0, 1.0)) / 2.0;
        new_solution[j] = opt->best_solution.position[j] + perturbation;
    }
    enforce_bound_constraints(opt);
}

// Accept or reject the new solution based on fitness and temperature
void accept_solution(Optimizer *opt, double *new_solution, double new_fitness, double temperature) {
    double delta_fitness = new_fitness - opt->best_solution.fitness;
    if (delta_fitness < 0 || rand_double(0.0, 1.0) < exp(-delta_fitness / temperature)) {
        for (int j = 0; j < opt->dim; j++) {
            opt->best_solution.position[j] = new_solution[j];
        }
        opt->best_solution.fitness = new_fitness;
    }
}

// Update the best solution (if current is better)
void update_best_solution_teo(Optimizer *opt) {
    // Already handled in accept_solution for TEO
}

// Cool the temperature
void cool_temperature(double *temperature) {
    *temperature *= COOLING_RATE;
}

// Main Optimization Function
void TEO_optimize(Optimizer *opt, double (*objective_function)(double *)) {
    double *current_solution = (double *)malloc(opt->dim * sizeof(double));
    double current_fitness = INFINITY;
    double temperature = INITIAL_TEMPERATURE;

    // Initialize the solution
    initialize_solution(opt, current_solution, &current_fitness, objective_function);

    // Main optimization loop
    for (int iter = 0; iter < opt->max_iter; iter++) {
        // Generate and evaluate new solution
        double *new_solution = (double *)malloc(opt->dim * sizeof(double));
        perturb_solution(opt, new_solution);
        double new_fitness = objective_function(new_solution);

        // Accept or reject the new solution
        accept_solution(opt, new_solution, new_fitness, temperature);

        // Cool the temperature
        cool_temperature(&temperature);

        // Check for convergence
        if (temperature < FINAL_TEMPERATURE) {
            free(new_solution);
            break;
        }

        free(new_solution);
    }

    free(current_solution);
}
